python/src/fi_wrapper.cpp

```diff
@@ -73,37 +87,88 @@ void bind_fi_sketch(py::module &m, const char* name) {
+    using namespace datasketches;
+    clazz.def(
         "get_serialized_size_bytes",
-        [](const frequent_items_sketch<T>& sk) { return sk.get_serialized_size_bytes(); },
+        [](const frequent_items_sketch<T, W, H, E>& sk) { return sk.get_serialized_size_bytes(); },
         "Computes the size needed to serialize the current state of the sketch. This can be expensive since every item needs to be looked at."
     )
     .def(
         "serialize",
-        [](const frequent_items_sketch<T>& sk) {
+        [](const frequent_items_sketch<T, W, H, E>& sk) {
           auto bytes = sk.serialize();
           return py::bytes(reinterpret_cast<const char*>(bytes.data()), bytes.size());
         },
-        "Serializes the sketch into a bytes object"
+        "Serializes the sketch into a bytes object."
     )
     .def_static(
         "deserialize",
-        [](const std::string& bytes) { return frequent_items_sketch<T>::deserialize(bytes.data(), bytes.size()); },
+        [](const std::string& bytes) { return frequent_items_sketch<T, W, H, E>::deserialize(bytes.data(), bytes.size()); },
         py::arg("bytes"),
-        "Reads a bytes object and returns the corresponding frequent_strings_sketch"
+        "Reads a bytes object and returns the corresponding frequent_strings_sketch."
     );
 }
 
+// py::object or any other type that requires a provided serde
+template<typename T, typename W, typename H, typename E, typename std::enable_if<!std::is_arithmetic<T>::value && !std::is_same<std::string, T>::value, bool>::type>
+void add_serialization(py::class_<datasketches::frequent_items_sketch<T, W, H, E>>& clazz) {
+    using namespace datasketches;
+    clazz.def(
+        "get_serialized_size_bytes",
+        [](const frequent_items_sketch<T, W, H, E>& sk, py_object_serde& serde) { return sk.get_serialized_size_bytes(serde); },
+        py::arg("serde"),
+        "Computes the size needed to serialize the current state of the sketch using the provided serde. This can be expensive since every item needs to be looked at."
+    )
+    .def(
+        "serialize",
+        [](const frequent_items_sketch<T, W, H, E>& sk, py_object_serde& serde) {
+          auto bytes = sk.serialize(0, serde);
+          return py::bytes(reinterpret_cast<const char*>(bytes.data()), bytes.size());
+        }, py::arg("serde"),
+        "Serializes the sketch into a bytes object using the provided serde."
+    )
+    .def_static(
+        "deserialize",
+        [](const std::string& bytes, py_object_serde& serde) {
+          return frequent_items_sketch<T, W, H, E>::deserialize(bytes.data(), bytes.size(), serde);
+        }, py::arg("bytes"), py::arg("serde"),
+        "Reads a bytes object using the provided serde and returns the corresponding frequent_strings_sketch."
+    );
+}
+
+// calls class __hash__ method
+struct py_hash_caller {
+  size_t operator()(const py::object& a) {
```

Comment:
this can be const, not sure if really matters

Reply:
right, this is done now